balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -664,12 +700,60 @@ func (b *pickfirstBalancer) endFirstPassIfPossibleLocked(lastErr error) {
 	// Start re-connecting all the SubConns that are already in IDLE.
 	for _, v := range b.subConns.Values() {
 		sd := v.(*scData)
-		if sd.state == connectivity.Idle {
+		if sd.connectivityState == connectivity.Idle {
 			sd.subConn.Connect()
 		}
 	}
 }
 
+func (b *pickfirstBalancer) updateSubConnHealthState(sd *scData, state balancer.SubConnState) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	// Previously relevant SubConns can still callback with state updates.
+	// To prevent pickers from returning these obsolete SubConns, this logic
+	// is included to check if the current list of active SubConns includes
+	// this SubConn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	switch state.ConnectivityState {
+	case connectivity.Ready:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+	case connectivity.TransientFailure:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: fmt.Errorf("health check failure: %v", state.ConnectionError)},
+		})
+	case connectivity.Connecting:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
+	default:
+		b.logger.Errorf("Got unexpected health update for SubConn %p: %v", state)
+	}
+}
+
+func (b *pickfirstBalancer) updateConcludedStateLocked(newState balancer.State) {
```

Comment:
Docstring for this one please. And here and in the next method, also please clearly indicate that a new picker is sent. Because "update concluded state" could simply also mean updating the `concludedState` field. Thanks.

Reply:
Added and modified the godoc.